pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,207 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package infracluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	mapiv1beta1 "github.com/openshift/api/machine/v1beta1"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	azurev1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	errUnableToGetAzureMAPIProviderSpec = errors.New("unable to get Azure MAPI ProviderSpec")
+)
+
+const (
+	clusterSecretName               = "capz-manager-cluster-credential"    // #nosec G101
+	capzManagerBootstrapCredentials = "capz-manager-bootstrap-credentials" // #nosec G101
+)
+
+// ensureAzureCluster ensures the AzureCluster cluster object exists.
+//
+//nolint:funlen
+func (r *InfraClusterController) ensureAzureCluster(ctx context.Context, log logr.Logger) (client.Object, error) {
+	target := &azurev1.AzureCluster{ObjectMeta: metav1.ObjectMeta{
+		Name:      r.Infra.Status.InfrastructureName,
+		Namespace: defaultCAPINamespace,
+	}}
+
+	capzManagerBootstrapCredentialsKey := client.ObjectKey{Namespace: defaultCAPINamespace, Name: capzManagerBootstrapCredentials}
+	capzManagerBootstrapCredentials := &corev1.Secret{}
+
+	if err := r.Get(ctx, capzManagerBootstrapCredentialsKey, capzManagerBootstrapCredentials); err != nil {
+		return nil, fmt.Errorf("failed to create Azure Cluster Secret: %w", err)
+	}
+
+	// Checking whether InfraCluster object exists. If it doesn't, create it.
+	if err := r.Get(ctx, client.ObjectKeyFromObject(target), target); err != nil && !apierrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get InfraCluster: %w", err)
+	} else if err == nil {
+		return target, nil
+	}
+
+	log.Info(fmt.Sprintf("AzureCluster %s/%s does not exist, creating it", target.Namespace, target.Name))
+
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return nil, fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	azureClientID, ok := capzManagerBootstrapCredentials.Data["azure_client_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureClientID: %w", err)
+	}
+
+	azureTenantID, ok := capzManagerBootstrapCredentials.Data["azure_tenant_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureTenantID: %w", err)
+	}
+
+	providerSpec, err := getAzureMAPIProviderSpec(ctx, r.Client)
```

Comment:
For debugging, we might want to output this to the logs. The providerSpec.ResourceGroup is the value we're using for the CAPI AzureCluster's ResourceGroup, and based on https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-capi-operator/189/pull-ci-openshift-cluster-capi-operator-main-e2e-azure-capi-techpreview/1833880963109621760/artifacts/e2e-azure-capi-techpreview/gather-extra/artifacts/pods/openshift-cluster-api_cluster-capi-operator-6ff69b6f54-d66cc_cluster-capi-operator.log that value is empty.

Reply:
Thanks Nolan, I've done that now! I think you're right that it's empty at this point